core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala

```diff
@@ -179,10 +183,126 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments, repository
+                repository.queries.getEntity(workspaceId, entityType, entityName)
+              } map {
+                case Some(entityRec) => CountAndSource(1, Source.single(entityRec))
+                case None            => CountAndSource(0, Source.empty)
+              }
+            } else if (entityQuery.columnFilter.isDefined) {
+              val columnFilter = entityQuery.columnFilter.get
+              repository.dataSource
+                .inTransaction(ReadCommitted) { _ =>
+                  repository.queries.countEntitiesWithColumnFilter(workspaceId, entityType, columnFilter)
+                }
+                .map { count =>
+                  CountAndSource(
+                    count,
+                    streamQuery(count,
+                                repository.queries
+                                  .queryEntitiesWithColumnFilter(workspaceId, entityType, entityQuery, columnFilter)
+                    )
+                  )
+                }
+            } else {
+              // there is no filter so we can just use the unfiltered count
+              Future.successful(
+                CountAndSource(
+                  unfilteredCount,
+                  streamQuery(unfilteredCount,
+                              repository.queries.queryEntitiesWithNoFilter(workspaceId, entityType, entityQuery)
+                  )
+                )
+              )
+            }
+
+          filteredCountAndSourceF.map { filteredCountAndSource =>
+            val pageCount: Int = Math.ceil(filteredCountAndSource.count.toFloat / entityQuery.pageSize).toInt
+            if (filteredCountAndSource.count > 0 && entityQuery.page > pageCount) {
+              throw new DataEntityException(
+                code = StatusCodes.BadRequest,
+                message = s"requested page ${entityQuery.page} is greater than the number of pages $pageCount"
+              )
+            }
+            (EntityQueryResultMetadata(unfilteredCount, filteredCountAndSource.count, pageCount),
+             filteredCountAndSource.source.map(_.toEntity)
+            )
+          }
+        }
+      }
+  }
+
+  private def streamQuery(count: Int,
+                          query: SqlStreamingAction[Seq[CompactEntityRecord], CompactEntityRecord, Effect.Read]
+  ) = {
+    import repository.dataSource.dataAccess.driver.api._
+    if (count == 0) {
+      // if there are no results, we can just return an empty source
+      Source.empty
+    } else {
+      // otherwise, we need to stream the results
+      Source.fromPublisher(
+        repository.dataSource.database.stream(
+          query.transactionally
+            .withTransactionIsolation(ReadCommitted)
+            .withStatementParameters(rsType = ResultSetType.ForwardOnly,
+                                     rsConcurrency = ResultSetConcurrency.ReadOnly,
+                                     fetchSize = repository.dataSource.fetchSize
+            )
+        )
+      )
+    }
+  }
```

Comment:
I like how you lifted the SqlStreamingAction -> Source functionality up here to the provider - it makes testing the individual queries much easier